src/ir/auth_logic/ast.h

```diff
@@ -139,6 +139,55 @@ class Assertion {
   AssertionVariantType value_;
 };
 
+enum class ConstantType { kNumberType, kPrincipalType };
+
+// Relation declaration types are of 3 forms
+// 1. NumberType
+// 2. PrincipalType
+// 3. User defined CustomType (string to store name of the type)
+class ArgumentType {
+ public:
+  enum class Kind { kNumber, kPrincipal, kCustom };
+  explicit ArgumentType(Kind kind, std::string name)
+      : kind_(std::move(kind)), name_(std::move(name)) {}
+  const Kind& kind() const { return kind_; }
+  const std::string name() const { return name_; }
+
+ private:
+  Kind kind_;
+  std::string name_;
+};
+
+class Argument {
+ public:
+  explicit Argument(std::string argument_name, ArgumentType argument_type)
```

Comment:
optional: `std::string argument_name` could be `absl::string_view` as in  `ArgumentType` case.